### PR TITLE
codestyle: Type-annotate web_util.py

### PR DESCRIPTION
**Pull request by stefanberger**
Type-annotate web_util.py. The following type annotation forces the casting of rest_params["api_version"] to a str. We can do this since we know that "api_version" has been set in get_restful_params.

def get_restful_params(urlstring: str) -> Dict[str, Union[str, None]]:

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

### CODE REVIEW

File: keylime/web_util.py [hunk 1]
```diff
@@ -440,11 +449,11 @@ def validate_api_version(handler, version, logger):
...
 
 
-def _list_to_dict(alist):
+def _list_to_dict(alist: List[str]) -> Dict[str, Union[str, None]]:
     """Convert list into dictionary via grouping [k0,v0,k1,v1,...]"""
     params = {}
     i = 0
     while i < len(alist):
-        params[alist[i]] = alist[i + 1] if (i + 1) < len(alist) else None
+        params[alist[i]] = alist[i + 1] if (i + 1) < len(alist) else None  # Must keep NONE here
```
Comment (stefanberger) [inline keylime/web_util.py:457]: It's unfortunate that there is this `None` here. An empty string "" **may be** equivalent an would help to make the return parameter a `Dict[str, str]` but then we do have issues with things like these here (in registrar_common.py for example):

```
        rest_params = web_util.get_restful_params(self.path)

        [...]

        agent_id = rest_params["agents"]

        if agent_id is not None:
```

File: keylime/web_util.py [hunk 2]
```diff
@@ -406,14 +415,14 @@ def echo_json_response(handler, code, status=None, results=None):
...
-def get_restful_params(urlstring):
+def get_restful_params(urlstring: str) -> Dict[str, Union[str, None]]:
     """Returns a dictionary of paired RESTful URI parameters"""
     parsed_path = urllib.parse.urlsplit(urlstring.strip("/"))
     query_params = urllib.parse.parse_qsl(parsed_path.query)
     path_tokens = parsed_path.path.split("/")
 
     # If first token looks like an API version, validate it and make sure it's supported
-    api_version = 0
+    api_version = "0"
```
Comment (stefanberger) [inline keylime/web_util.py:425]: This avoids `Dict[str, Union[str, int, None]]`

Comment (stefanberger) [reply keylime/web_util.py:425]: I had to adjust the test case because of this change.

File: keylime/web_util.py [hunk 3]
```diff
@@ -406,14 +415,14 @@ def echo_json_response(handler, code, status=None, results=None):
     return False
 
 
-def get_restful_params(urlstring):
+def get_restful_params(urlstring: str) -> Dict[str, Union[str, None]]:
```
Comment (aplanas) [inline keylime/web_util.py:418]: Not sure, but maybe `Dict[str, Optional[str]]`

File: keylime/web_util.py (hunk 1 above)
Comment (aplanas) [reply keylime/web_util.py:457]: I personally never do a `if var is [not] None`, but `if [not] var`. Maybe this helps here as for one side the code will be equivalent and the type can be forced.

File: keylime/web_util.py (hunk 3 above)
Comment (stefanberger) [reply keylime/web_util.py:418]: right. I did this kind of on purpose in this case to make the `None` visible.

File: keylime/web_util.py (hunk 1 above)
Comment (stefanberger) [reply keylime/web_util.py:457]: I would put it on a todo list because I think this kind of change requires a bit more 'care'.

**Comment by stefanberger** [comment]
> @stefanberger once the annotations pass also mypy you can add the following to the mypy.ini:
> 
> ```
> [mypy-keylime.web_util]
> ignore_errors = False
> ```

That's only going to be possible once other low level functions have been annotated.


```
mypy run-test: commands[0] | mypy keylime/
keylime/web_util.py:41: error: Call to untyped function "get_secdir" in typed context  [no-untyped-call]
keylime/web_util.py:86: error: Call to untyped function "setpassword" in typed context  [no-untyped-call]
keylime/web_util.py:99: error: Call to untyped function "cmd_init" in typed context  [no-untyped-call]
keylime/web_util.py:118: error: Call to untyped function "cmd_mkcert" in typed context  [no-untyped-call]
keylime/web_util.py:133: error: Call to untyped function "cmd_mkcert" in typed context  [no-untyped-call]
keylime/web_util.py:280: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
keylime/web_util.py:294: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
keylime/web_util.py:310: error: Incompatible types in assignment (expression has type "None", variable has type "str")  [assignment]
keylime/web_util.py:400: error: Incompatible types in assignment (expression has type "bytes", variable has type "str")  [assignment]
keylime/web_util.py:406: error: Argument 1 to "write" of "IO" has incompatible type "str"; expected "bytes"  [arg-type]
```

**Review by aplanas — APPROVED** [review]

**Review by THS-on — APPROVED** [review]

**Pull request merged**